modules/cnn_3dobj/include/opencv2/cnn_3dobj.hpp

```diff
@@ -0,0 +1,252 @@
+         */
+        CV_WRAP static void writeBinaryfile(string filenameImg, const char* binaryPath, const char* headerPath, int num_item, int label_class, int x, int y, int z, int isrgb);
+    };
+
+/** @brief Caffe based 3D images descriptor.
+ The class extract feature which is discriminant for class and pose from given samples @cite wohlhart15.
+ */
+
+/************************************ Feature Extraction Class ************************************/
+    class CV_EXPORTS_W descriptorExtractor
+    {
+        private:
+        caffe::Net<float>* convnet;
+        cv::Size input_geometry;
+        int num_channels;
+        bool net_set;
+        int net_ready;
+        cv::Mat mean_;
+        std::vector<string> device_info;
+        /** @brief Load the mean file in binaryproto format if it is needed.
+        @param mean_file Path of mean file which stores the mean of training images, it is usually generated by Caffe tool.
+         */
+        void setMean(const string& mean_file);
+        /** @brief Wrap the input layer of the network in separate cv::Mat objects(one per channel).
+         This way we save one memcpy operation and we don't need to rely on cudaMemcpy2D.
+         The last preprocessing operation will write the separate channels directly to the input layer.
+         */
+        void wrapInput(std::vector<cv::Mat>* input_channels);
+        /** @brief Convert the input image to the input image format of the network.
+         */
+        void preprocess(const cv::Mat& img, std::vector<cv::Mat>* input_channels);
+        public:
+        /** @brief Set the device for feature extraction, if the GPU is used, there should be a device_id.
+        @param device_type CPU or GPU.
+        @param device_id ID of GPU.
+         */
+        descriptorExtractor(const string& device_type, int device_id);
+        /** @brief Get device information for feature extraction.
+         */
+        std::vector<string> getDevice();
+        /** @brief Set device information for feature extraction, if the descriptor is already loaded and
+         you want to change the device for featrue extraction, this function is convenient.
+        @param device_type CPU or GPU.
+        @param device_id ID of GPU.
+         */
+        void setDevice(const string& device_type, const string& device_id = "");
+        /** @brief Initiate a classification structure, the net work parameter is stored in model_file,
+         the network structure is stored in trained_file, you can decide whether to use mean images or not.
+        @param model_file Path of caffemodel which including all parameters in CNN.
+        @param trained_file Path of prototxt which defining the structure of CNN.
+        @param mean_file Path of mean file(option).
+         */
+        void loadNet(const string& model_file, const string& trained_file, string mean_file = "");
+        /** @brief Extract features from a single image or a set of images which is collected in a vector.
```

Comment:
replace:
Extract features from a single image or from a vector of images. If loadNet was not called before, this method invocation will fail